pkg/controller/nodeimage/nodeimage_controller.go

```diff
@@ -0,0 +1,479 @@
+		err = r.Get(context.TODO(), types.NamespacedName{Name: name}, nodeImage)
+		if err != nil {
+			return err
+		}
+
+		modified, messages, tmpDuration = r.doUpdateNodeImage(nodeImage, node)
+		if !modified {
+			return nil
+		}
+		return r.Update(context.TODO(), nodeImage)
+	})
+	duration.Merge(tmpDuration)
+	if err != nil {
+		klog.Errorf("Failed to update NodeImage %s spec(%v): %v", name, messages, err)
+	} else if modified {
+		klog.Infof("Successfully update NodeImage %s spec(%v)", name, messages)
+	}
+	return modified, err
+}
+
+func (r *ReconcileNodeImage) doUpdateNodeImage(nodeImage *appsv1alpha1.NodeImage, node *v1.Node) (modified bool, messages []string, wait *requeueduration.Duration) {
+	wait = &requeueduration.Duration{}
+	if node != nil {
+		if !reflect.DeepEqual(nodeImage.Labels, node.Labels) {
+			modified = true
+			messages = append(messages, "node labels changed")
+			nodeImage.Labels = node.Labels
+		}
+	}
+
+	newImageMap := make(map[string]appsv1alpha1.ImageSpec, len(nodeImage.Spec.Images))
+	for name, imageSpec := range nodeImage.Spec.Images {
+		var newTags []appsv1alpha1.ImageTagSpec
+		for i := range imageSpec.Tags {
+			tagSpec := &imageSpec.Tags[i]
+			fullName := fmt.Sprintf("%s:%s", name, tagSpec.Tag)
+
+			// If tag has owners which have been deleted, delete this tag
+			if len(tagSpec.OwnerReferences) > 0 {
+				var activeRefs []v1.ObjectReference
+				for _, ref := range tagSpec.OwnerReferences {
+					gvk := ref.GroupVersionKind()
+					if gvk.Group != controllerKind.Group || gvk.Kind != "ImagePullJob" {
+						activeRefs = append(activeRefs, ref)
+						break
+					}
+					job := appsv1alpha1.ImagePullJob{}
+					err := r.Get(context.TODO(), types.NamespacedName{Namespace: ref.Namespace, Name: ref.Name}, &job)
+					if err != nil {
+						if errors.IsNotFound(err) {
+							continue
+						}
+						klog.Errorf("Failed to check owners for %s in NodeImage %s, get job %v error: %v", fullName, name, util.DumpJSON(ref), err)
+						activeRefs = append(activeRefs, ref)
+						break
+					}
+					if job.UID != ref.UID {
+						klog.Warningf("When check owners for %s in NodeImage %s, get job %v find UID %v not equal", fullName, name, util.DumpJSON(ref), job.UID)
+						continue
+					}
+					activeRefs = append(activeRefs, ref)
+					break
+				}
+				if len(activeRefs) != len(tagSpec.OwnerReferences) {
+					modified = true
+					messages = append(messages, fmt.Sprintf("image %s owners changed from %v to %v",
+						fullName, util.DumpJSON(tagSpec.OwnerReferences), util.DumpJSON(activeRefs)))
+					if len(activeRefs) == 0 {
+						continue
+					}
+					tagSpec.OwnerReferences = activeRefs
+				}
+			}
+
+			// If tag has TTL and status has completed, prepare to delete this tag
+			if tagSpec.CreatedAt != nil && tagSpec.PullPolicy != nil && tagSpec.PullPolicy.TTLSecondsAfterFinished != nil {
+
+				var completionTime *metav1.Time
+				if imageStatus := nodeImage.Status.ImageStatuses[name]; len(imageStatus.Tags) > 0 {
```

Comment:
can we skip the length check to reduce the code indentation? It seems that it works just the same.